service/src/main/java/com/codeforcommunity/processor/LeaderboardProcessorImpl.java

```diff
@@ -0,0 +1,64 @@
+package com.codeforcommunity.processor;
+
+import static org.jooq.generated.Tables.*;
+import static org.jooq.generated.tables.Users.USERS;
+import static org.jooq.impl.DSL.count;
+
+import com.codeforcommunity.api.ILeaderboardProcessor;
+import com.codeforcommunity.dto.leaderboard.GetLeaderboardRequest;
+import com.codeforcommunity.dto.leaderboard.GetLeaderboardResponse;
+import com.codeforcommunity.dto.leaderboard.LeaderboardEntry;
+import com.codeforcommunity.enums.ReservationAction;
+import java.util.List;
+import org.jooq.DSLContext;
+
+public class LeaderboardProcessorImpl implements ILeaderboardProcessor {
+  private final DSLContext db;
+
+  public LeaderboardProcessorImpl(DSLContext db) {
+    this.db = db;
+  }
+
+  @Override
+  public GetLeaderboardResponse getUsersLeaderboard(
+      GetLeaderboardRequest getUsersLeaderboardRequest) {
+    // select user_id, username and number of blocks completed
+    List<LeaderboardEntry> users =
+        db.select(USERS.ID, USERS.USERNAME, count())
+            .from(USERS)
+            .leftJoin(RESERVATIONS)
+            .on(USERS.ID.eq(RESERVATIONS.USER_ID))
```

Comment:
use `.onKey()` whenever possible, it'll just automatically use the fk relationship between the two tables